### PR TITLE
Web linter

### PR DESCRIPTION
**Pull request by mugling**
Follows discussion in #18477

> Sure, sounds fine, and mod_cgi would be simplest from my point of view

Extends linter to support being run as a CGI script. Tries very hard to avoid consuming unnecessary resources by performing client-side JSON validation and only sending a content body if linting was required.

@narc0tiq can be installed using the following:

```
Alias /tools /path/to/cataclysm/tools
<Directory "/path/to/cataclysm/tools/format">
    DirectoryIndex format.html
    Options +ExecCGI

    <Files "format.pl">
       LimitRequestBody 524288
       <LimitExcept POST>
            Order allow,deny
            Deny from all
        </LimitExcept>
       SetEnv PATH /usr/local/bin
       SetHandler cgi-script
    </Files>

    Require all granted
</Directory>
```

The only presumption is that you have `perl` installed in `/usr/local/bin` and that you also have the `CGI` module. It does **not** depend on `jq`

The HTML5 web interface is intentionally simple and has no jQuery bloat or other indeed any other dependencies at all. If anyone knows a lightweight JSON syntax highlighter that could go in as it might be helpful for new users getting used to editing the game data.

### CODE REVIEW

**Comment by illi-kun** [comment]
Looking forward on this!

**Comment by mugling** [comment]
> Looking forward on this!

It's tempting to bundle this with one of the better HTML5 editors - syntax highlighting is always helpful but especially for new contributors. Separate PR as for now the priority is to get this running.

**Comment by cainiaowu** [comment]
Is it possible for linter to check grammar issue? Like double space after comma that kind of things, or even typos. Typos could be mark as warnings.

**Comment by DangerNoodle** [comment]
Will be interesting to see. So will will this be used, and how will new contributors find out what to do to use it?

**Comment by mugling** [comment]
> Is it possible for linter to check grammar issue? Like double space after comma that kind of things, or even typos. Typos could be mark as warnings.

I did think of that but the false positive rate would be huge unless carefully designed

> So will will this be used, and how will new contributors find out what to do to use it?

Presuming you meant how will this be used you will go to a link online, paste your JSON and click 'lint' and that will be it

**Comment by DangerNoodle** [comment]
That actually sounds infinitely simpler than what I was expecting.

**Comment by mugling** [comment]
> That actually sounds infinitely simpler than what I was expecting.

I can't think of a simpler design. I might later add syntax highlighting and I'm not averse to other suggestions. Soon the linter will support flags (possibly with hints as to the different possibilities).

We are steadily increasing towards half a million lines of JSON so to take no action would be negligent.

**Comment by DangerNoodle** [comment]
That seems perfect in fact, thank you. My assumptions were that future plans would have been a tool either run in command line, one run in Git shell, or an embedded tool that can be used drag-and-drop like with the AStyle tool. All of those I could reasonably figure out how to use, but might trip up a someone less experience.

In contrast, this idea sounds as easy to use as it will ever get.

**Comment by mugling** [comment]
> In contrast, this idea sounds as easy to use as it will ever get.

Rule 15 of the internet applies

File: tools/format/format.html [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+                  elem.value = xhr.responseText;
+              } else if (xhr.status == 304) {
+                  // no-op
+              } else if (xhr.status == 400) {
+                  alert(xhr.responseText);
+              } else {
+                  alert('Linter currently unavailable');
+              }
+          };
+          xhr.send(encodeURI('data=' + elem.value));
```
Comment (narc0tiq) [inline tools/format/format.html]: This needs to be more like `xhr.send('data=' + encodeURIComponent(elem.value));` -- any of `, / ? : @ & = + $ #` should be URL-encoded, which is what `encodeURIComponent` does (`encodeURI` leaves them alone).

**Comment by narc0tiq** [comment]
Ok, for testing I've set up the script as of 3bcd54f (plus the change in my comment) at <http://dev.narc.ro/cgi-bin/format/format.pl>, with the front-end at <http://dev.narc.ro/cataclysm/format.html> (being fed off some symlinks from my `~/src`; intention is to have it run from Jenkins' reference repo eventually).

Seems to be working nicely so far. You guys should try to break it a bit more before I make it fully live.

**Comment by DangerNoodle** [comment]
Will be interesting to test this out.

**Comment by mugling** [comment]
> Will be interesting to test this out.

Then do!

**Comment by DangerNoodle** [comment]
Interesting so far, presently testing the old recipe_ammo.json.

Already I am seeing that the messages being reported do not show the "example of canonical format" information that the commit-checker has been set to present.

**Comment by mugling** [comment]
@DangerNoodle elaborate?

**Comment by DangerNoodle** [comment]
Wait. No, this is incorrect. I forgot that "unmatch contex" is actually a gap in the linter contex list, not a linting failure.

**Comment by mugling** [comment]
It can also be you defined an umatched field. Until we have 100% coverage the message remains ambiguous.

**Comment by DangerNoodle** [comment]
True, but in this instance I simply pasted the contents of recipe_ammo, and the missing contex being cited is the basic "recipe" contex.

What files have the linter currently been built to examine?

**Comment by mugling** [comment]
@DangerNoodle what is the full context? Coverage is supposed to be 100% for `recipe`

**Comment by DangerNoodle** [comment]
I posted a screenshot of the error message above. Context is:

`Unmatched contex 'recipe<recipe>:note'`

I deleted the only instance of a recipe comment using "note" instead of "//" and it appears to be taking longer now, I assume it is working through the actual formatting failures.

**Comment by DangerNoodle** [comment]
There we go. It finished fixing the errors without generating any message indicating it finished working.

And yes, using `note` for a comment was apparently the problem. This was not on my part though, as it was simply copy-paste from Windows graphical build 5581.

**Comment by mugling** [comment]
@narcotiq do you have the perl JSON XS module (eg. the C compiled version, it's about 25% faster than pure perl). The next step up performance wise is of course `mod_perl` but there is no evidence that we need this at present.

**Comment by mugling** [comment]
Latest commit prevents submitting a new linting request until the previous has completed - the intention is to reduce workload on the server from processing duplicate requests. A busy cursor is now displayed to make it clear that a request is pending.

**Comment by narc0tiq** [comment]
Ok, I've updated to d11f2fd and applied a variant of the snippet from the OP to make it work without any changes -- looks good to me. Once this is merged I can target it to the Jenkins reference repo and have it continue working and updating automatically for the foreseeable future.

**Comment by narc0tiq** [comment]
Ack, wrong button.

**Comment by DangerNoodle** [comment]
> Latest commit prevents submitting a new linting request until the previous has completed - the intention is to reduce workload on the server from processing duplicate requests. A busy cursor is now displayed to make it clear that a request is pending.

Excellent.

**Comment by mugling** [comment]
> Once this is merged I can target it to the Jenkins reference repo and have it continue working and updating automatically for the foreseeable future.

I've applied your local fix to use `encodeComponentURI` and merged this to `master`.

I'm going to target performance improvements in the next iteration. We might start with client side compression initially as uploading ~200KB per linting cycle could be slow for users on an asynchronous connection. @narc0tiq could we discuss out-of-band and exchange some log files so I can get an idea of where to target performance optimizations. I now have a forum account if you want to PM me.

**Comment by DangerNoodle** [comment]
A pleasure to see this in use then.

**Comment by narc0tiq** [comment]
@mugling I'm on the IRC, but I don't think we're on the same schedule (likely timezone issues). I'll toss you my email through a forum PM and we'll go from there.

I've updated the live one to use the reference copy Jenkins does; this might not be ideal, as I'm not sure if it's still available while a build is running. Regardless, the canonical URL is either http://dev.narc.ro/cataclysm/format.html or http://dev.narc.ro/tools/format -- I tend to prefer the former, but they are both supposed to work regardless.

**Comment by DangerNoodle** [comment]
Now to ensure that the link will be provided somewhere that it will actually be seen by others to be used.

I have also noticed that trying to run the entire undivided recipe files tends to make it time out. Presumably more reason to split up recipes in the way others have been doing.

**Comment by mugling** [comment]
> I have also noticed that trying to run the entire undivided recipe files tends to make it time out

We could bump the timeout. What is the **upload** speed of your connection?

**Comment by DangerNoodle** [comment]
Apologies, I had not thought of that. Given the poor quality of my connection (use of satellite net), this concern should perhaps be disregarded.

Testing states it is 4.36 Mbps, and I do not recall what my ISP claims it should be.

Regardless, there appears to be no issue with formatting any of the split-up recipe files so far.

**Comment by pisskop** [comment]
The web linter is returning a no error, but jenkins is rejecting the files based on the order of the lines.  will the be synchronized?

**Comment by narc0tiq** [comment]
Should be literally the same code. No error doesn't mean no changes,
though. Try pasting from the web linter into another file and comparing
with your original.

**Comment by mugling** [comment]
> No error doesn't mean no changes, though.

I would suspect probably this although feel free to post an example if this is not the case

**Pull request merged**